src/Shared/Dictionary/AdaptiveCapacityDictionary.cs

```diff
@@ -0,0 +1,752 @@
+        /// <see cref="IDictionary{TKey, TValue}"/> or <see cref="IReadOnlyDictionary{TKey, TValue}"/>
+        /// or an object with public properties as key-value pairs.
+        /// </param>
+        /// <param name="comparer">Equality comparison.</param>
+        /// <param name="capacity">Initial capacity.</param>
+        public AdaptiveCapacityDictionary(IEnumerable<KeyValuePair<TKey, TValue>> values, int capacity, IEqualityComparer<TKey> comparer)
+        {
+            _comparer = comparer ?? EqualityComparer<TKey>.Default;
+
+            _arrayStorage = new KeyValuePair<TKey, TValue>[capacity];
+
+            foreach (var kvp in values)
+            {
+                Add(kvp.Key, kvp.Value);
+            }
+        }
+
+        /// <inheritdoc />
+        public TValue this[TKey key]
+        {
+            get
+            {
+                if (key == null)
+                {
+                    ThrowArgumentNullExceptionForKey();
+                }
+
+                TryGetValue(key, out var value);
+
+                return value!;
+            }
+
+            set
+            {
+                if (key == null)
+                {
+                    ThrowArgumentNullExceptionForKey();
+                }
+
+                if (_dictionaryStorage != null)
+                {
+                    _dictionaryStorage[key] = value;
+                    return;
+                }
+
+                var index = FindIndex(key);
+                if (index < 0)
+                {
+                    EnsureCapacity(_count + 1);
+                    if (_dictionaryStorage != null)
+                    {
+                        _dictionaryStorage[key] = value;
+                        return;
+                    }
+                    _arrayStorage[_count++] = new KeyValuePair<TKey, TValue>(key, value);
+                }
+                else
+                {
+                    _arrayStorage[index] = new KeyValuePair<TKey, TValue>(key, value);
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public int Count
+        {
+            get
+            {
+                if (_dictionaryStorage != null)
+                {
+                    return _dictionaryStorage.Count;
+                }
+                return _count;
+            }
+        }
+
+        /// <inheritdoc />
+        public IEqualityComparer<TKey> Comparer
+        {
+            get
+            {
+                return _comparer ?? EqualityComparer<TKey>.Default;
+            }
+        }
+
+        /// <inheritdoc />
+        bool ICollection<KeyValuePair<TKey, TValue>>.IsReadOnly => false;
+
+        /// <inheritdoc />
+        public ICollection<TKey> Keys
+        {
+            get
+            {
+                if (_dictionaryStorage != null)
+                {
+                    return _dictionaryStorage.Keys;
+                }
+
+                var array = _arrayStorage;
```

Comment:
We should add a comment about if we think this operation is comment and a potential future optimization could be to store a keys and values array and return those here without a copy.

Reply:
That would be spicy. I don't think it's that common though.